src/node_quic_socket.h

```diff
@@ -88,12 +90,170 @@ class JSQuicSocketListener : public QuicSocketListener {
+class QuicPacket : public MemoryRetainer {
+ public:
+  // Creates a new QuicPacket. By default the packet will be
+  // stack allocated with a max size of NGTCP2_MAX_PKTLEN_IPV4.
+  // If a larger packet size is specified, it will be heap
+  // allocated. Generally speaking, a QUIC packet should never
+  // be larger than the current MTU to avoid IP fragmentation.
+  //
+  // The content of a QuicPacket is provided by ngtcp2. The
+  // typical use pattern is to create a QuicPacket instance
+  // and then pass a pointer to it's internal buffer and max
+  // size in to an ngtcp2 function that serializes the data.
+  // ngtcp2 will fill the buffer as much as possible then return
+  // the number of bytes serialized. User code is then responsible
+  // for calling SetLength() to set the final length of the
+  // QuicPacket prior to sending it off to the QuicSocket.
+  //
+  // The diagnostic label is used in NODE_DEBUG_NATIVE output
+  // to differentiate send operations. This should always be
+  // a statically allocated string or nullptr (in which case
+  // the value "unspecified" is used in the debug output).
+  //
+  // Instances of std::unique_ptr<QuicPacket> are moved through
+  // QuicSocket and ultimately become the responsibility of the
+  // SendWrap instance. When the SendWrap is cleaned up, the
+  // QuicPacket instance will be freed.
+  static std::unique_ptr<QuicPacket> Create(
+      const char* diagnostic_label = nullptr,
+      size_t len = NGTCP2_MAX_PKTLEN_IPV4) {
+    return std::make_unique<QuicPacket>(diagnostic_label, len);
+  }
+
+  // Copy the data of the QuicPacket to a new one. Currently,
+  // this is only used when retransmitting close connection
+  // packets from a QuicServer.
+  static std::unique_ptr<QuicPacket> Copy(
+      const std::unique_ptr<QuicPacket>& other) {
+    return std::make_unique<QuicPacket>(*other.get());
+  }
+
+  QuicPacket(const char* diagnostic_label, size_t len) :
+      data_(len),
+      diagnostic_label_(diagnostic_label) {
+    CHECK_LE(len, NGTCP2_MAX_PKT_SIZE);
+  }
+
+  QuicPacket(const QuicPacket& other) {
+    diagnostic_label_ = other.diagnostic_label_;
+    data_.AllocateSufficientStorage(other.data_.length());
+    memcpy(*data_, *other.data_, other.data_.length());
+  }
+
+  uint8_t* data() { return *data_; }
+  size_t length() const { return data_.length(); }
+  void SetLength(size_t len) {
+    CHECK_LE(len, data_.length());
+    data_.SetLength(len);
+  }
+
+  const char* diagnostic_label() const {
+    return diagnostic_label_ != nullptr ?
+        diagnostic_label_ : "unspecified";
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("data", data_.length());
+    tracker->TrackFieldWithSize("label", strlen(diagnostic_label_));
+  }
+
+  SET_MEMORY_INFO_NAME(QuicPacket);
+  SET_SELF_SIZE(QuicPacket);
+
+ private:
+  MaybeStackBuffer<uint8_t, NGTCP2_MAX_PKTLEN_IPV4> data_;
```

Comment:
This means we always allocate the max packet length, even if the data packet is smaller (which I assume is the common case)… is that okay?

Reply:
Currently, it's not yet clear what the most common case will be. We need to do some testing and see what a reasonable default max is. We can adjust this in the future once we have a better idea